README.md
# Nerves.Firmware

Elixir API for upgrading and managing firmware on a Nerves device.

**Looking for over-the-network firmware updates?** see
[nerves_firmware_http](https://github.com/nerves-project/nerves_firmware_http), which provides an HTTP micro-service providing over-network firmware management.

Leans heavily on Frank Hunleth's excellent [fwup](https://github.com/fhunleth/fwup), which is included of the standard Nerves configurations.

For more, read the [documentation](https://hexdocs.pm/nerves_firmware).

## Installation

It's published in [Hex](https://hex.pm/nerves_firmware), so..

  1. Add nerves_firmware to your list of dependencies in `mix.exs`:

          def deps do
            [{:nerves_firmware, "~> 0.4.0"}]
          end

  2. Ensure nerves_firmware is started before your application:

          def application do
            [applications: [:nerves_firmware]]
          end

## Usage
in `config.exs` you can configure the block device and signing of the firmware.

``` elixir
use Mix.Config

config :nerves_firmware,
  device: "/dev/mmcblk0",
  pub_key_path: "/etc/fwup-key.pub"
```


CHANGELOG.md
# Changelog

## v0.4.0 (2017-05-04)

- Support chunked streaming direct to fwup (thanks @mobileoverlord and @fhunleth)
- Support fwup's signing process for firmware (thanks @connorrigby)
- Add finalize/1 (mostly to support `nerves_firmware_http` streaming state)

## v0.3.0 (2017-02-22)

- Converted compile-time config to runtime
- Fixed tests to work with newer fwup (>= 0.11.0)
- bumped to elixir 1.4 & fixed warnings

## v0.2.0 (2016-06-23)

- Split nerves_firmware_http out, depends on this module
- Cleaned up documentation somewhat


test/test_helper.exs
ExUnit.start()

defmodule Helpers do

  def firmware_file(filename) do
    Path.join "test/firmware_files", filename
  end
end

test/nerves_firmware_test.exs
defmodule Nerves.Firmware.Test do
  @moduledoc false
  use ExUnit.Case
  doctest Nerves.Firmware
  require Logger
  alias Nerves.Firmware

  @app_id :nerves_firmware

  test "Firmware app started succesfully" do
    assert Application.start(@app_id) == {:error, {:already_started, @app_id}}
  end

  test "returning proper status before and after firmware upgrade" do
    device = Application.get_env(:nerves_firmware, :device, "/tmp/test_firmware.fw")
    fw = firmware_file("test_1.fw")
    assert :ok == Firmware.Fwup.apply(fw, device, :complete)
    assert Firmware.state()[:status] == :active
    #now, try installing firmware upgrade
    metrics1 = read_firmware_metrics(device)
    assert metrics1 == read_firmware_metrics(device)
    #assert Firmware.apply(fw, :upgrade) == :ok
    assert Firmware.upgrade_and_finalize(fw) == :ok
    metrics2 = read_firmware_metrics(device)
    assert metrics1 != metrics2
    # now that we've update firmware, we should be in await_restart state
    assert Firmware.state()[:status] == :await_restart
    #assert Firmware.apply(fw, :upgrade) == :error
    assert Firmware.upgrade_and_finalize(fw) == {:error, :await_restart}
    metrics3 = read_firmware_metrics(device)
    # and firmware should not be updated
    assert metrics2 == metrics3
    assert metrics1 != metrics3
  end

  # HELPER FUNCTIONS

  defp read_firmware_metrics(device) do
    {read_mbr(device)}
  end

  defp read_mbr(device) do
    File.open device, [:read], fn(file) ->
      IO.binread(file, 512)
    end
  end

  defp firmware_file(filename) do
    Path.join "test/firmware_files", filename
  end

end


test/nerves_firmware_cowboy_test.exs
defmodule Nerves.Firmware.Adapters.Cowboy.Test do

  use ExUnit.Case, async: true
  import Helpers

  @device Application.get_env(:nerves_firmware, :device)

  defmodule TestServer do
    @moduledoc "configure cowboy to handle firmware on port 8088"

    @test_http_port 8088
    @test_http_path "/fw_http_test"
    @test_http_uri Path.join("localhost:#{@test_http_port}", @test_http_path)

    def uri, do: @test_http_uri

    def start do
      :ok = Application.start :ranch
      :ok = Application.start :cowlib
      :ok = Application.start :cowboy
      dispatch = :cowboy_router.compile [ {:_, [
        {@test_http_path, Nerves.Firmware.Adapters.Cowboy, []}
      ]} ]
      :cowboy.start_http(:http, 10, [port: @test_http_port], [env: [dispatch: dispatch]])
    end
  end

  {:ok, _} = HTTPotion.start
  {:ok, _} = TestServer.start

  test "returning status and installing firmware upgrade" do
    fw = firmware_file("test_1.fw")
    # create the low level firmware file
    assert :ok == Nerves.Firmware.Fwup.apply(fw, @device, :complete)    

    # now, test the firmware
    assert get_firmware_state[:status] == "active"
    #now, try installing firmware upgrade
   
    s = File.stat fw
    # delay 1500ms to force different mtime for updated firmware
    :timer.sleep 1500 
    assert 204 = send_firmware(fw)
    # REVIEW: SuperLame test
    s2 = File.stat(fw)
    assert s != s2
    # now that we've update firmware, we should be in await_restart state
    assert get_firmware_state[:status] == "await_restart"
    # this should fail with 403 since firmware is not yet rebooted
    assert 403 = send_firmware(fw)
    # and firmware should not be updated
    assert s != File.stat(fw)
  end

  defp get_firmware_state do
    resp = HTTPotion.get TestServer.uri, headers: ["Accept": "application/json"]
    assert resp.status_code == 200
    json_to_term(resp)
  end
  
  defp send_firmware(path) do
    resp = HTTPotion.put(TestServer.uri, [
           body: File.read!(path),
           headers: ["Content-Type": "application/x-firmware"]])
    resp.status_code
  end

  defp json_to_term(resp) do
    content_type = Keyword.fetch(resp.headers.hdrs, :'content-type')
    assert {:ok, "application/json"} == content_type
    {:ok, term} = JSX.decode(resp.body, [{:labels, :atom}])
    Enum.into term, []
  end

end


mix.exs
defmodule Nerves.Firmware.Mixfile do
  use Mix.Project

  @version "0.4.0"

  def project do
    [
      app: :nerves_firmware,
      version: @version,
      elixir: "~> 1.4",
      build_embedded: Mix.env == :prod,
      start_permanent: Mix.env == :prod,
      deps: deps(Mix.env),
      description: "Manage and update firmware on a Nerves device",
      package: package(),
      name: "Nerves.Firmware",
      docs: docs()  ]
  end

  def application do
    [ applications: [:logger], mod: {Nerves.Firmware, []} ]
  end

  defp deps(_) do
    [{:ex_doc, "~> 0.15", only: :dev}]
  end

  defp docs do
    [
      source_ref: "v#{@version}",
      main: "Nerves.Firmware",
      source_url: "https://github.com/nerves-project/nerves_firmware",
      extras: [ "README.md", "CHANGELOG.md"]
    ]
  end

  defp package do
    [ maintainers: ["Garth Hitchens", "Justin Schneck"],
      licenses: ["Apache-2.0"],
      links: %{github: "https://github.com/nerves-project/nerves_firmware"}]
  end
end


lib/firmware.ex
defmodule Nerves.Firmware do
  @moduledoc """
  API for upgrading and managing firmware on a Nerves device.

  Handles firmware for a single block device (like /dev/mmcblk0). Delegates a
  lot to Frank Hunleth's excellent [fwup](https://github.com/fhunleth/fwup).

  Provides:
  - Firmware upgrades
  - Firmware status
  - Firmware-related activities (shutdown, reboot, halt)

  **Looking for over-the-network firmware updates?** see
  [nerves_firmware_http](https://github.com/nerves-project/nerves_firmware_http),
  which provides an HTTP micro-service providing over-network firmware management.

  ## Installation

    1. Add nerves_firmware to your list of dependencies in `mix.exs`:

          def deps do
            [{:nerves_firmware, "~> 0.3.0"}]
          end

    2. Ensure nerves_firmware is started before your application:

          def application do
            [applications: [:nerves_firmware]]
          end

  ## Configuration

  In your app's config.exs, you can configure the block device for your
  target that is managed by setting the device key as follows:

        config :nerves_firmware, device: "dev/mmcblk0"
  """

  use Application
  require Logger

  @type reason :: atom

  @typedoc """
  Arguments to be passed to FWUP.
  """
  @type args :: [binary]

  @server Nerves.Firmware.Server

  @doc """
  Application start callback.
  """
  @spec start(atom, term) :: {:ok, pid} | {:error, String.t}
  def start(_type, _args) do
    import Supervisor.Spec, warn: false
    Logger.debug "#{__MODULE__}.start(...)"
    opts = [strategy: :one_for_one, name: Nerves.Firmware.Supervisor]
    children = [ worker(@server, []) ]
    Supervisor.start_link(children, opts)
  end

  @doc """
  Return a map of information about the current firmware.

  This currently contains values showing the state of the firmware installation,
  as well as the key/value pairs encoded in the firmware itself.

  __status:__

  `:active` - Currently running the latest firmware received.  Firmware
  must be in this state to be updated.

  `:await_restart` - Firmware has been updated since restart, and a restart is
  needed to start running from the new firmware.

  __device:__

  The device file that holds the firmware, e.g. /dev/mmcblk0
  """
  @spec state() :: Map.t
  def state(), do: GenServer.call @server, {:state}

  @doc """
  Applies a firmware file to the device media.

  This mostly just passes information through to Nerves.Firmware.Fwup.apply(..)
  which is a very thin wrapper around [fwup](https://github.com/fhunleth/fwup), but it
  also sets the firwmare state based on the action to reflect the update, and
  prevent multiple updates from overwriting known good firmware.

  * `action` can be one of `:upgrade` or `:complete`
  * `args` is a list of extra arguments to be passed to fwup.

  Returns {:error, :await_restart} if the upgrade is requested after
  already updating an image without a reboot in-between.
  """
  @spec apply(String.t, atom, args) :: :ok | {:error, reason}
  def apply(firmware, action, args \\ []) do
    args = maybe_pub_key_args(args)
    GenServer.call @server, {:apply, firmware, action, args}
  end

  @doc """
  Returns `true` if new firmware can currently be installed.

  The firmware module usually allows new firmware to be installed, but there
  are situations where installing new firmware is dangerous.  Currently
  if the device has had an update applied without being restarted,
  this returns false, and update apis will return errors, to prevent bricking.
  """
  @spec allow_upgrade?() :: true | false
  def allow_upgrade?() do
    GenServer.call @server, {:allow_upgrade?}
  end

  @doc """
  Apply a 1 or 2-phase nerves update

  Applies firmware using `upgrade` task, then, if /tmp/finalize.fw exists,
  apply that file with `on-reboot` task.  Supports @fhunleth 2-phase format.

  * `args` is a list of extra arguments to be passed to fwup.

  Returns {:error, :await_restart} if the upgrade is requested after
  already updating an image without a reboot in-between.
  """
  @spec upgrade_and_finalize(String.t, args) :: :ok | {:error, reason}
  def upgrade_and_finalize(firmware, args \\ []) do
    args = maybe_pub_key_args(args)
    GenServer.call @server, {:upgrade_and_finalize, firmware, args}, :infinity
  end

  @doc """
  Applies /tmp/finalize.fw if with `on-reboot` task if exists,

  * `args` is a list of extra arguments to be passed to fwup, but is currently
  ignored for this function.

  Returns {:error, :await_restart} if the finalize is requested after
  already updating an image without a reboot in-between.
  """
  @spec finalize(args) :: :ok | {:error, reason}
  def finalize(args \\ []) do
    args = maybe_pub_key_args(args)
    # REVIEW args is ignored by the server for this call. What should they do?
    GenServer.call @server, {:finalize, args}, :infinity
  end

  @doc """
  Reboot the device.

  Issues the os-level `reboot` command, which reboots the device, even
  if erlinit.conf specifies not to reboot on exit of the Erlang VM.
  """
  @spec reboot() :: :ok
  def reboot(), do: logged_shutdown "reboot"

  @doc """
  Reboot the device gracefully

  Issues :init.stop command to gracefully shutdown all applications in the Erlang VM.
  All code is unloaded and ports closed before the system terminates by calling halt(Status).
  erlinit.config must be set to reboot on exit(default) for a graceful reboot to work.
  """
  @spec reboot(atom) :: :ok
  def reboot(:graceful), do: :init.stop

  @doc """
  Forces device to power off (without reboot).
  """
  @spec poweroff() :: :ok
  def poweroff(), do: logged_shutdown "poweroff"

  @doc """
  Forces device to halt (meaning hang, not power off, nor reboot).

  Note: this is different than :erlang.halt(), which exists BEAM, and
  may end up rebooting the device if erlinit.conf settings allow reboot on exit.
  """
  @spec halt() :: :ok
  def halt(), do: logged_shutdown "halt"

  # private helpers

  defp logged_shutdown(cmd, args \\ []) do
    Logger.info "#{__MODULE__} : device told to #{cmd}"
    System.cmd(cmd, args)
    :ok
  end

  @spec maybe_pub_key_args(args) :: args
  defp maybe_pub_key_args(args) do
    pub_key_path = Application.get_env(:nerves_firmware, :pub_key_path)
    if pub_key_path do
      Logger.info "#{__MODULE__} using signature"
      ["-p", "#{pub_key_path}" | args]
    else
      args
    end
  end
end


lib/firmware/server.ex
defmodule Nerves.Firmware.Server do
  @moduledoc false

  use GenServer
  alias Nerves.Firmware.Fwup
  require Logger

  @type reason :: term
  @type state :: Struct.t

  defmodule State do
    @moduledoc false
    defstruct status: :active, device: nil
  end

  def start_link() do
    GenServer.start __MODULE__, :ok, name: __MODULE__
  end

  @spec init(term) :: {:ok, state} | {:error, reason}
  def init(_arg) do
    Logger.debug "#{__MODULE__}.init"
    device = Application.get_env(:nerves_firmware, :device, "/dev/mmcblk0")
    {:ok, %State{status: :active, device: device}}
  end

  def handle_call({:state}, _from, state) do
    {:reply, public_state(state), state}
  end

  def handle_call({:allow_upgrade?}, _from, state) do
    {:reply, allow_upgrade?(state), state}
  end

  def handle_call({:apply, firmware, action, args}, _from, state) do
    try_apply_if_allowed state, fn() ->
      Fwup.apply(firmware, state.device, action, args)
    end
  end

  def handle_call({:upgrade_and_finalize, firmware, args}, _from, state) do
    try_apply_if_allowed state, fn() ->
      do_upgrade_and_finalize(firmware, args, state)
    end
  end

  def handle_call({:finalize, args}, _from, state) do
    try_apply_if_allowed state, fn() ->
      do_finalize(args, state)
    end
  end

  defp try_apply_if_allowed(state, afn) do
    if allow_upgrade?(state) do
      try_apply(state, afn)
    else
      {:reply, {:error, :await_restart}, state}
    end
  end

  defp try_apply(state, afn) do
    case afn.() do
      :ok ->
        {:reply, :ok, %{state | status: :await_restart}}
      {:error, reason} ->
        {:reply, {:error, reason}, state}
    end
  end

  defp allow_upgrade?(state) do
    state.status == :active
  end

  @spec do_upgrade_and_finalize(String.t, [binary], Struct.t) :: :ok | {:error, reason}
  defp do_upgrade_and_finalize(firmware, args, state) do
    finalize_fw = Application.get_env(:nerves_firmware, :finalize_fw, "/tmp/finalize.fw")
    File.rm(finalize_fw)
    case Fwup.apply(firmware, state.device, "upgrade", args) do
      {:error, reason} ->
        Logger.error "#{__MODULE__} upgrade failed: #{inspect reason}"
        {:error, reason}
      :ok ->
        Logger.info "upgrade succeded"
        do_finalize(args, state)
    end
  end

  def do_finalize(_args, state) do
    finalize_fw = Application.get_env(:nerves_firmware, :finalize_fw, "/tmp/finalize.fw")
    if File.exists?(finalize_fw) do
      Logger.info "Found #{finalize_fw}, applying finalize/on-reboot"
      try_finalize(finalize_fw, state)
    else
      :ok
    end
  end

  # called after upgrade to see if we have a finalize.fw for a 2-phase
  # firmware update.  If so, apply it by running the on-reboot task.
  @spec try_finalize(String.t, state) :: :ok | {:error, :reason}
  defp try_finalize(ffw, state) do
    case Fwup.apply(ffw, state.device, "on-reboot", []) do
      :ok ->
        Logger.info "finalize/on-reboot succeded, ready for restart"
        :ok
      {:error, reason} ->
        Logger.info "finalize/on-reboot failed: #{inspect reason}"
        {:error, reason}
    end
  end

  # return the public "state" of the firmware (not genserver state)
  defp public_state(state) do
    %{status: state.status, device: state.device}
  end

end


lib/nerves_firmware/fwup.ex
defmodule Nerves.Firmware.Fwup do
  use GenServer
  require Logger

  @moduledoc """
  A Port interface to stream firmware to fwup

  Example usage
    file = File.read!(/path/to/my.fw)
    {:ok, pid} = Nerves.Firmware.Fwup.start_link([device: "/tmp/test.img", task: "complete"])
    Nerves.Firmware.Fwup.stream_chunk(pid, file, await: true)
    Nerves.Firmware.Fwup.stop

  """

  @timeout 120_000

  def start_link(opts \\ []) do
    opts = Keyword.put_new(opts, :callback, self())
    GenServer.start_link(__MODULE__, opts)
  end

  def stop(pid) do
    GenServer.stop(pid)
  end

  def stream_chunk(pid, chunk, opts \\ [await: false]) do
    Logger.debug "Sending Chunk: #{inspect chunk}"
    GenServer.call(pid, {:stream_chunk, chunk, opts}, @timeout)
  end

  def init(opts) do
    Process.flag(:trap_exit, true)
    device = opts[:device] || Application.get_env(:nerves_firmware, :device, "/dev/mmcblk0")
    task = opts[:task] || "upgrade"
    fwup = System.find_executable("fwup")
    callback = opts[:callback]
    port = Port.open({:spawn_executable, fwup},
      [{:args, ["-aFU", "-d", device, "-t", task]},
        {:packet, 4},
        :use_stdio,
        :binary,
        :exit_status])
    {:ok, %{
      port: port,
      byte_size: 0,
      callback: callback
    }}
  end

  def handle_call({:stream_chunk, chunk, opts}, from, s) do
    Port.command(s.port, chunk)
    Logger.debug "Chunk Sent"
    #send s.port, {self(), {:command, chunk}}
    case opts[:await] do
      true -> {:noreply, %{s | callback: from}}
      false -> {:reply, :ok, s}
    end
  end

  def handle_info({_port, {:data, <<"OK", _code :: integer-16>>}}, s) do
    Logger.debug "FWUP Done"
    GenServer.reply(s.callback, :ok)
    {:noreply, s}
  end

  def handle_info({_port, {:data, <<"ER", code :: integer-16, message :: binary>>}}, s) do
    Logger.debug "FWUP Error #{code}: #{message}"
    {:noreply, s}
  end

  def handle_info({_port, {:data, <<warning :: binary-2, code :: integer-16, message :: binary>>}}, s)
    when warning in ["WA", "WN"] do
    Logger.debug "FWUP Warning #{code}: #{message}"
    {:noreply, s}
  end

  def handle_info({_port, {:data, <<"PR", progress :: integer-16>>}}, s) do
    Logger.debug "FWUP Progress: #{progress}%"
    {:noreply, s}
  end

  def handle_info({_port, {:data, resp}}, s) do
    Logger.debug "FWUP unknown response: #{inspect resp}"
    {:noreply, s}
  end

  def handle_info(_msg, s) do
    {:noreply, s}
  end

  @doc """
  Apply the firmware in <input> to the given <device>, executing <task>.

  `args` is a list of arguments to be passed to fwup.

  Not implemented using ports, because ports cant send EOF, so it's not possible
  to stream firmware through a port.  Porcelain doesn't work because `goon` isn't
  easy to compile for the target in Nerves.

  The simple file-based I/O allows using named pipes to solve the streaming issues.
  """
  @spec apply(String.t, String.t, String.t, [binary]) :: :ok | {:error, term}
  def apply(input, device, task, args \\ []) do
    Logger.info "Firmware: applying #{task} to #{device}"
    fwup_args =
      ["-aqU", "--no-eject", "-i", input, "-d", device, "-t", task] ++ args
    case System.cmd("fwup", fwup_args) do
      {_out, 0} ->
        :ok
      {error, _} ->
        Logger.error error
        {:error, :fwup_error}
    end
  end
end


lib/firmware/adapters/cowboy.ex
defmodule Nerves.Firmware.Adapters.Cowboy do
  @moduledoc """
  Implements a Cowboy HTTP/REST adapter for Nerves.Firmware

  Defines an _acceptor_ that receives and installs firmware updates. Simply use
  use __HTTP PUT__ to send a firmware to the URI for the device, specifying
  Content-Type `application/x-firmware`

  Also defines a _provider_ for the `application/json` content type that allows
  an HTTP GET to reutrn information on the current firmware status and metadata
  in JSON form.

  ## Configuration Example

  Reference the adapter in your cowboy dispatch routes wherever you want.
  To place the resource at "/firmware", for instance, do:

  ```elixir
  cowboy_dispatch = :cowboy_router.compile [ {:_, [
    { "/firmware", Nerves.Firmware.Adapters.Cowboy, [] }
  ]} ]
  ```
  """

  require Logger

  @max_upload_chunk 100000        # 100K max chunks to keep memory reasonable
  @max_upload_size  100000000     # 100M max file to avoid using all of flash
  @tmp_fw_path "/tmp/uploaded.fw"

  def init(_transport, _req, _state) do
    {:upgrade, :protocol, :cowboy_rest}
  end

  def rest_init(req, handler_opts) do
    {:ok, req, handler_opts}
  end

  def allowed_methods(req, state) do
    {["GET", "PUT", "POST"], req, state}
  end

  def content_types_provided(req, state) do
    {[
      {"application/json", :json_provider}
    ], req, state}
  end

  def content_types_accepted(req, state) do
    {[
      {{"application", "x-firmware", []}, :upload_acceptor}
    ], req, state}
  end

  def json_provider(req, state) do
    term = Nerves.Firmware.state
    {:ok, body} = JSX.encode(term, [{:space, 1}, {:indent, 2}])
    { body <> "\n", req, state}
  end

  @doc """
  Acceptor for cowboy to update firmware via HTTP.

  Once firmware is streamed, it returns success (2XX) or failure (4XX/5XX).
  Calls `update_status()` to reflect status at `/sys/firmware`.
  Won't let you upload firmware on top of provisional (returns 403)
  """
  def upload_acceptor(req, state) do
		Logger.info "request to receive firmware"
    if Nerves.Firmware.allow_upgrade? do
      upload_and_apply_firmware(req, state)
    else
      {:halt, reply_with(403, req), state}
		end
  end

  # TODO:  Ideally we'd like to allow streaming directly to fwup, but its hard
  # due to limitations with ports and writing to fifo's from elixir
  # Right solution would be to get Porcelain fixed to avoid golang for goon.
  defp upload_and_apply_firmware(req, state) do
		Logger.info "receiving firmware"
		File.open!(@tmp_fw_path, [:write], &(stream_fw &1, req))
    Logger.info "firmware received"
    response = case Nerves.Firmware.apply(@tmp_fw_path, :upgrade) do
      {:error, _whatever} ->
        {:halt, reply_with(400, req), state}
      {:ok, _fw_metadata} ->  # TODO: consider returning metadata
        {true, req, state}
      :ok ->
        {true, req, state}
    end
    File.rm @tmp_fw_path
    response
  end

  # helper to return errors to requests from cowboy more easily
  defp reply_with(code, req) do
    {:ok, req} = :cowboy_req.reply(code, [], req)
    req
  end

  # copy from a cowboy req into a IO.Stream
  defp stream_fw(f, req, count \\ 0) do
    #  send an event about (bytes_uploaded: count)
    if count > @max_upload_size do
      {:error, :too_large}
    else
      case :cowboy_req.body(req, [:length, @max_upload_chunk]) do
        {:more, chunk, new_req} ->
          :ok = IO.binwrite f, chunk
          stream_fw(f, new_req, (count + byte_size(chunk)))
        {:ok, chunk, new_req} ->
          :ok = IO.binwrite f, chunk
          {:done, new_req}
      end
    end
  end
end


lib/firmware/http.ex
defmodule Nerves.Firmware.HTTP do
  @moduledoc """
  Implements a lightweight HTTP/REST service for Firmware

  Defines an _acceptor_ that receives and installs firmware updates. Simply use
  use __HTTP PUT__ to send a firmware to the URI for the device, specifying
  Content-Type `application/x-firmware`

  Also defines a _provider_ for the `application/json` content type that allows
  an HTTP GET to reutrn information on the current firmware status and metadata
  in JSON form.
  ```
  """

  require Logger

  @max_upload_chunk 100000        # 100K max chunks to keep memory reasonable
  @max_upload_size  100000000     # 100M max file to avoid using all of flash

  @http_port Application.get_env(:nerves_firmware, :http_port, 8988)
  @http_path Application.get_env(:nerves_firmware, :http_path, "/firmware")
  @upload_path Application.get_env(:nerves_firmware, :upload_path, "/tmp/uploaded.fw")

  def start do
    Logger.info "starting service #{__MODULE__} on #{@http_port}#{@http_path}"
    dispatch = :cowboy_router.compile [{:_,[{@http_path, __MODULE__, []}]}]
    :cowboy.start_http(__MODULE__, 10, [port: @http_port], [env: [dispatch: dispatch]])
  end

  def init(_transport, _req, _state) do
    {:upgrade, :protocol, :cowboy_rest}
  end

  def rest_init(req, handler_opts) do
    {:ok, req, handler_opts}
  end

  def allowed_methods(req, state) do
    {["GET", "PUT", "POST"], req, state}
  end

  def content_types_provided(req, state) do
    {[
      {"application/json", :json_provider}
    ], req, state}
  end

  def content_types_accepted(req, state) do
    {[
      {{"application", "x-firmware", []}, :upload_acceptor}
    ], req, state}
  end

  def json_provider(req, state) do
    term = Nerves.Firmware.state
    {:ok, body} = JSX.encode(term, [{:space, 1}, {:indent, 2}])
    { body <> "\n", req, state}
  end

  @doc """
  Acceptor for cowboy to update firmware via HTTP.

  Once firmware is streamed, it returns success (2XX) or failure (4XX/5XX).
  Calls `update_status()` to reflect status at `/sys/firmware`.
  Won't let you upload firmware on top of provisional (returns 403)
  """
  def upload_acceptor(req, state) do
		Logger.info "request to receive firmware"
    if Nerves.Firmware.allow_upgrade? do
      upload_and_apply_firmware_upgrade(req, state)
    else
      {:halt, reply_with(403, req), state}
		end
  end

  # TODO:  Ideally we'd like to allow streaming directly to fwup, but its hard
  # due to limitations with ports and writing to fifo's from elixir
  # Right solution would be to get Porcelain fixed to avoid golang for goon.
  defp upload_and_apply_firmware_upgrade(req, state) do
		Logger.info "receiving firmware"
		File.open!(@upload_path, [:write], &(stream_fw &1, req))
    Logger.info "firmware received"
    response = case Nerves.Firmware.upgrade_and_finalize(@upload_path) do
      {:error, _reason} ->
        {:halt, reply_with(400, req), state}
      :ok ->
        case :cowboy_req.header("x-reboot", req) do
          {:undefined, _} ->  nil
          {_, _} -> Nerves.Firmware.reboot
        end
        {true, req, state}
    end
    File.rm @upload_path
    response
  end

  # helper to return errors to requests from cowboy more easily
  defp reply_with(code, req) do
    {:ok, req} = :cowboy_req.reply(code, [], req)
    req
  end

  # copy from a cowboy req into a IO.Stream
  defp stream_fw(f, req, count \\ 0) do
    #  send an event about (bytes_uploaded: count)
    if count > @max_upload_size do
      {:error, :too_large}
    else
      case :cowboy_req.body(req, [:length, @max_upload_chunk]) do
        {:more, chunk, new_req} ->
          :ok = IO.binwrite f, chunk
          stream_fw(f, new_req, (count + byte_size(chunk)))
        {:ok, chunk, new_req} ->
          :ok = IO.binwrite f, chunk
          {:done, new_req}
      end
    end
  end
end
